File: native/src/volume.c
#include "teeforge.h"
#include <unistd.h>
#include <fcntl.h>
#include <linux/input.h>
#include <stdio.h>
#include <time.h>

/* ===== 音量键监听 Volume Key Listener ===== */
/* 独立模块，方便复用 Standalone module for reuse */

/* 监听音量键 Listen for volume key */
/* timeout_sec: 超时秒数 Timeout in seconds */
/* 返回: REGION_CN / REGION_GLOBAL / -1(超时) */
int volume_listen(int timeout_sec) {
    const char *devices[] = {
        "/dev/input/event0", "/dev/input/event1",
        "/dev/input/event2", "/dev/input/event3",
        "/dev/input/event4", "/dev/input/event5",
        NULL
    };

    int fds[16];
    int nfds = 0;

    /* 打开所有输入设备 Open all input devices */
    for (int i = 0; devices[i] != NULL && nfds < 16; i++) {
        int fd = open(devices[i], O_RDONLY | O_NONBLOCK);
        if (fd >= 0) {
            fds[nfds++] = fd;
        }
    }

    if (nfds == 0) {
        log_msg(LOG_WARN, "无法打开输入设备 [Cannot open input devices]");
        return -1;
    }

    time_t start = time(NULL);
    struct input_event ev;

    while (time(NULL) - start < timeout_sec) {
        for (int i = 0; i < nfds; i++) {
            ssize_t n = read(fds[i], &ev, sizeof(ev));
            if (n == sizeof(ev)) {
                if (ev.type == EV_KEY && ev.value == 1) {
                    if (ev.code == KEY_VOLUMEUP) {
                        for (int j = 0; j < nfds; j++) close(fds[j]);
                        return REGION_CN;
                    } else if (ev.code == KEY_VOLUMEDOWN) {
                        for (int j = 0; j < nfds; j++) close(fds[j]);
                        return REGION_GLOBAL;
                    }
                }
            }
        }
        usleep(50000);
    }

    for (int j = 0; j < nfds; j++) close(fds[j]);
    return -1;
}

File: crates/teeforge/src/volume.rs
use crate::error::{Result, TfError};
#[cfg(target_os = "android")]
use std::ffi::CString;
#[cfg(target_os = "android")]
use std::fs;
#[cfg(target_os = "android")]
use std::mem::MaybeUninit;
#[cfg(target_os = "android")]
use std::os::fd::RawFd;
#[cfg(target_os = "android")]
use std::time::{Duration, Instant};

#[cfg(target_os = "android")]
#[repr(C)]
struct InputEvent {
    time: libc::timeval,
    event_type: u16,
    code: u16,
    value: i32,
}

#[cfg(target_os = "android")]
const EV_KEY: u16 = 1;
#[cfg(target_os = "android")]
const KEY_VOLUMEUP: u16 = 115;
#[cfg(target_os = "android")]
const KEY_VOLUMEDOWN: u16 = 114;

#[cfg(target_os = "android")]
fn open_events() -> Result<Vec<RawFd>> {
    let mut entries = fs::read_dir("/dev/input")?
        .filter_map(std::result::Result::ok)
        .filter(|entry| entry.file_name().to_string_lossy().starts_with("event"))
        .collect::<Vec<_>>();
    entries.sort_by_key(|entry| entry.file_name());
    let mut fds = Vec::new();
    for entry in entries {
        let path = entry.path();
        let Ok(c_path) = CString::new(path.as_os_str().as_encoded_bytes()) else {
            continue;
        };
        // SAFETY: c_path is NUL-terminated and the returned descriptor is owned here.
        let fd = unsafe { libc::open(c_path.as_ptr(), libc::O_RDONLY | libc::O_NONBLOCK) };
        if fd >= 0 {
            fds.push(fd);
        }
    }
    if fds.is_empty() {
        Err(TfError::new("无法打开输入设备 [Cannot open input devices]"))
    } else {
        Ok(fds)
    }
}

#[cfg(target_os = "android")]
pub(crate) fn listen(timeout_seconds: u64) -> Result<i32> {
    let fds = open_events()?;
    let mut poll_fds = fds
        .iter()
        .map(|fd| libc::pollfd {
            fd: *fd,
            events: libc::POLLIN,
            revents: 0,
        })
        .collect::<Vec<_>>();
    let deadline = Instant::now() + Duration::from_secs(timeout_seconds);
    let result = 'waiting: loop {
        let remaining = deadline.saturating_duration_since(Instant::now());
        if remaining.is_zero() {
            break Ok(-1);
        }
        let timeout = remaining.as_millis().min(i32::MAX as u128) as i32;
        // SAFETY: poll_fds points to initialized pollfd values for the supplied length.
        let ready = unsafe { libc::poll(poll_fds.as_mut_ptr(), poll_fds.len() as _, timeout) };
        if ready < 0 {
            break Err(TfError::from(std::io::Error::last_os_error()));
        }
        for poll_fd in &poll_fds {
            if poll_fd.revents & libc::POLLIN == 0 {
                continue;
            }
            let mut event = MaybeUninit::<InputEvent>::uninit();
            // SAFETY: event points to writable storage of exactly InputEvent size.
            let read = unsafe {
                libc::read(
                    poll_fd.fd,
                    event.as_mut_ptr().cast(),
                    std::mem::size_of::<InputEvent>(),
                )
            };
            if read as usize != std::mem::size_of::<InputEvent>() {
                continue;
            }
            // SAFETY: read initialized every byte of event after the exact-size check.
            let event = unsafe { event.assume_init() };
            if event.event_type == EV_KEY && event.value == 1 {
                if event.code == KEY_VOLUMEUP {
                    break 'waiting Ok(1);
                }
                if event.code == KEY_VOLUMEDOWN {
                    break 'waiting Ok(0);
                }
            }
        }
    };
    for fd in fds {
        // SAFETY: every descriptor was returned by open_events and is closed exactly once.
        unsafe { libc::close(fd) };
    }
    result
}

#[cfg(not(target_os = "android"))]
pub(crate) fn listen(_timeout_seconds: u64) -> Result<i32> {
    Err(TfError::new(
        "音量监听仅支持 Android [Volume listening is Android-only]",
    ))
}
